File: backend/app/services/msg91.py
import httpx

from app.config import settings


def format_mobile(phone: str) -> str:
    digits = "".join(c for c in phone if c.isdigit())
    cc = settings.msg91_country_code
    if len(digits) == 10:
        return f"{cc}{digits}"
    if len(digits) == 12 and digits.startswith(cc):
        return digits
    return f"{cc}{digits}" if len(digits) <= 10 else digits


async def send_otp_sms(phone: str, otp: str) -> tuple[bool, str]:
    if not settings.msg91_auth_key or not settings.msg91_template_id:
        return False, "MSG91 not configured"

    mobile = format_mobile(phone)
    headers = {"Content-Type": "application/json", "authkey": settings.msg91_auth_key}

    # Template: "Your OTP is ##var1##. Do not share it with anyone."
    otp_payload: dict = {
        "template_id": settings.msg91_template_id,
        "mobile": mobile,
        "var1": otp,
        "otp": otp,
    }
    if settings.msg91_sender_id:
        otp_payload["sender"] = settings.msg91_sender_id

    async with httpx.AsyncClient(timeout=15.0) as client:
        resp = await client.post(
            "https://control.msg91.com/api/v5/otp",
            headers=headers,
            json=otp_payload,
        )
        if resp.is_success and "success" in resp.text.lower():
            return True, "OTP sent via SMS"

        flow_payload: dict = {
            "template_id": settings.msg91_template_id,
            "short_url": "0",
            "recipients": [{"mobiles": mobile, "var1": otp}],
        }
        if settings.msg91_sender_id:
            flow_payload["sender"] = settings.msg91_sender_id

        flow = await client.post(
            "https://control.msg91.com/api/v5/flow/",
            headers=headers,
            json=flow_payload,
        )
        if flow.is_success and "success" in flow.text.lower():
            return True, "OTP sent via SMS"

        params = {
            "authkey": settings.msg91_auth_key,
            "mobile": mobile,
            "otp": otp,
            "otp_expiry": str(settings.otp_expire_minutes),
            "message": f"Your OTP is {otp}. Do not share it with anyone.",
        }
        if settings.msg91_sender_id:
            params["sender"] = settings.msg91_sender_id

        legacy = await client.get("https://api.msg91.com/api/sendotp.php", params=params)
        if legacy.is_success and "success" in legacy.text.lower():
            return True, "OTP sent via SMS"

        return False, flow.text or resp.text or "MSG91 send failed"

File: src/lib/msg91.ts
export type Msg91Config = {
	authKey: string;
	templateId: string;
	senderId?: string;
	countryCode: string;
};

export function getMsg91Config(env: Env): Msg91Config | null {
	const e = env as Env & {
		MSG91_AUTH_KEY?: string;
		MSG91_TEMPLATE_ID?: string;
		MSG91_SENDER_ID?: string;
		MSG91_COUNTRY_CODE?: string;
	};
	if (!e.MSG91_AUTH_KEY?.trim() || !e.MSG91_TEMPLATE_ID?.trim()) {
		return null;
	}
	return {
		authKey: e.MSG91_AUTH_KEY.trim(),
		templateId: e.MSG91_TEMPLATE_ID.trim(),
		senderId: e.MSG91_SENDER_ID?.trim(),
		countryCode: e.MSG91_COUNTRY_CODE ?? "91",
	};
}

export function isDevMode(env: Env): boolean {
	return (env as Env & { DEV_MODE?: string }).DEV_MODE === "true";
}

/** Indian 10-digit → 919876543210 */
export function formatMobile(phone: string, countryCode: string): string {
	const digits = phone.replace(/\D/g, "");
	if (digits.length === 10) return `${countryCode}${digits}`;
	if (digits.length === 12 && digits.startsWith(countryCode)) return digits;
	if (digits.length > 10) return digits;
	return `${countryCode}${digits}`;
}

function parseSuccess(status: number, text: string): boolean {
	const lower = text.toLowerCase();
	if (status < 200 || status >= 300) return false;
	try {
		const data = JSON.parse(text) as { type?: string };
		if (data.type === "error" || data.type === "fail") return false;
	} catch {
		/* not json */
	}
	return lower.includes("success") || status === 200;
}

type AttemptResult = { ok: boolean; detail: string };

async function tryRequest(
	label: string,
	url: string,
	init: RequestInit,
): Promise<AttemptResult> {
	try {
		const response = await fetch(url, init);
		const text = await response.text();
		if (parseSuccess(response.status, text)) {
			return { ok: true, detail: `${label}: ${text.slice(0, 120)}` };
		}
		return { ok: false, detail: `${label} (${response.status}): ${text.slice(0, 200)}` };
	} catch (err) {
		const msg = err instanceof Error ? err.message : String(err);
		return { ok: false, detail: `${label} (network): ${msg}` };
	}
}

/** Template: "Your OTP is ##var1##. Do not share it with anyone." */
export async function sendOtpSms(
	config: Msg91Config,
	phone: string,
	otp: string,
): Promise<{ ok: boolean; message: string; attempts: string[] }> {
	const mobile = formatMobile(phone, config.countryCode);
	const headers = {
		"Content-Type": "application/json",
		authkey: config.authKey,
	};
	const attempts: string[] = [];
	const id = config.templateId;
	const sender = config.senderId ?? "msg91";

	// Flow API (DLT templates with ##var1##)
	const flowRecipients = {
		mobiles: mobile,
		var1: otp,
		VAR1: otp,
	};
	const flowBodies = [
		{ flow_id: id, sender, recipients: [flowRecipients] },
		{ template_id: id, sender, short_url: "0", recipients: [flowRecipients] },
		{ flow_id: id, template_id: id, sender, recipients: [flowRecipients] },
	];

	for (let i = 0; i < flowBodies.length; i++) {
		const r = await tryRequest(
			`flow-v5-${i + 1}`,
			"https://control.msg91.com/api/v5/flow",
			{ method: "POST", headers, body: JSON.stringify(flowBodies[i]) },
		);
		attempts.push(r.detail);
		if (r.ok) return { ok: true, message: "OTP sent via SMS", attempts };
	}

	// OTP API v5
	const otpBodies = [
		{ template_id: id, mobile, var1: otp, otp, sender },
		{ template_id: id, mobile, OTP: otp, sender },
	];
	for (let i = 0; i < otpBodies.length; i++) {
		const r = await tryRequest(
			`otp-v5-${i + 1}`,
			"https://control.msg91.com/api/v5/otp",
			{ method: "POST", headers, body: JSON.stringify(otpBodies[i]) },
		);
		attempts.push(r.detail);
		if (r.ok) return { ok: true, message: "OTP sent via SMS", attempts };
	}

	// Legacy GET
	const params = new URLSearchParams({
		authkey: config.authKey,
		mobile,
		otp,
		otp_expiry: "10",
		otp_length: String(otp.length),
		DLT_TE_ID: id,
		message: `Your OTP is ${otp}. Do not share it with anyone.`,
		sender,
	});
	const legacy = await tryRequest(
		"legacy-sendotp",
		`https://api.msg91.com/api/sendotp.php?${params}`,
		{ method: "GET" },
	);
	attempts.push(legacy.detail);
	if (legacy.ok) return { ok: true, message: "OTP sent via SMS", attempts };

	return {
		ok: false,
		message: attempts[attempts.length - 1] ?? "MSG91 send failed",
		attempts,
	};
}

export function smsConfigStatus(env: Env) {
	const config = getMsg91Config(env);
	return {
		sms_configured: !!config,
		auth_key_set: !!(env as Env & { MSG91_AUTH_KEY?: string }).MSG91_AUTH_KEY?.trim(),
		template_id: (env as Env & { MSG91_TEMPLATE_ID?: string }).MSG91_TEMPLATE_ID ?? null,
		sender_id: (env as Env & { MSG91_SENDER_ID?: string }).MSG91_SENDER_ID ?? null,
		country_code: (env as Env & { MSG91_COUNTRY_CODE?: string }).MSG91_COUNTRY_CODE ?? "91",
		dev_mode: isDevMode(env),
	};
}
